aiven/resource_kafka_schema.go

```diff
@@ -0,0 +1,208 @@
+		Required:    true,
+		ForceNew:    true,
+	},
+	"service_name": {
+		Type:        schema.TypeString,
+		Description: "Service to link the Kafka Schema to",
+		Required:    true,
+		ForceNew:    true,
+	},
+	"subject_name": {
+		Type:        schema.TypeString,
+		Description: "Kafka Schema Subject name",
+		Required:    true,
+		ForceNew:    true,
+	},
+	"schema": {
+		Type:             schema.TypeString,
+		Description:      "Kafka Schema configuration should be a valid Avro Schema JSON format",
+		Required:         true,
+		ValidateFunc:     validation.ValidateJsonString,
+		StateFunc:        normalizeJsonString,
+		DiffSuppressFunc: diffSuppressJsonObject,
+	},
+	"version": {
+		Type:        schema.TypeInt,
+		Description: "Kafka Schema configuration version",
+		Computed:    true,
+	},
+}
+
+// diffSuppressJsonObject checks logical equivalences in JSON Kafka Schema values
+func diffSuppressJsonObject(k, old, new string, d *schema.ResourceData) bool {
+	var objOld, objNew interface{}
+
+	if err := json.Unmarshal([]byte(old), &objOld); err != nil {
+		return false
+	}
+	if err := json.Unmarshal([]byte(new), &objNew); err != nil {
+		return false
+	}
+
+	return reflect.DeepEqual(objNew, objOld)
+}
+
+// normalizeJsonString returns normalized JSON string
+func normalizeJsonString(v interface{}) string {
+	jsonString, _ := structure.NormalizeJsonString(v)
+
+	return jsonString
+}
+
+func resourceKafkaSchema() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKafkaSchemaCreate,
+		Update: resourceKafkaSchemaUpdate,
+		Read:   resourceKafkaSchemaRead,
+		Delete: resourceKafkaSchemaDelete,
+		Exists: resourceKafkaSchemaExists,
+		Importer: &schema.ResourceImporter{
+			State: resourceKafkaSchemaState,
+		},
+
+		Schema: aivenKafkaSchemaSchema,
+	}
+}
+
+func kafkaSchemaSubjectGetLastVersion(m interface{}, project, serviceName, subjectName string) (int, error) {
+	client := m.(*aiven.Client)
+
+	r, err := client.KafkaSubjectSchemas.GetVersions(project, serviceName, subjectName)
+	if err != nil {
+		return 0, err
+	}
+
+	var latestVersion int
+	for _, v := range r.Versions {
+		if v > latestVersion {
+			latestVersion = v
+		}
+	}
+
+	return latestVersion, nil
+}
+
+// Aiven Kafka schema creates a new Kafka Schema subject with a new version, and if
+// Kafka Schema subject with a given name already exists API will validate new Kafka
+// Schema configuration against latest previous version for compatibility and it
+// compatible will create a new version for the same Kafka Schema name
+func resourceKafkaSchemaCreate(d *schema.ResourceData, m interface{}) error {
+	project := d.Get("project").(string)
+	serviceName := d.Get("service_name").(string)
+	subjectName := d.Get("subject_name").(string)
+
+	client := m.(*aiven.Client)
+
+	// Add method retrieves back subject ID that we cannot use in Terraform composite
+	// ID since it is not used anywhere else, we need to know a newly created version
```

Comment:
This comment is hard to understand.

Reply:
indeed, will rewrite it

Reply:
this comment is no longer relevant, in the earlier implementation I was using version as a part of Terraform resource Id instead of Kafka Schema Subject ID. But right we are not using either version or Kafka Schema Subject ID for Terraform Resource ID ```	d.SetId(buildResourceID(project, serviceName, subjectName))```